File: ios/Classes/Native/NativeAdLoader.swift
//
//  NativeAdLoader.swift
//  nend_plugin
//
//

import UIKit
import NendAd

class NativeAdLoader: AdBridger {
    var onConnect: ((NADNative, String) -> Void)?
    
    private let client : NADNativeClient
    private var loadCompletion : NADNativeCompletionBlock!
    
    init(with param: AdUnitCodable, channel: FlutterMethodChannel) {
        client = NADNativeClient(spotId: String(param.adUnit.spotId), apiKey: param.adUnit.apiKey)
        super.init(with: channel, tag: .nativeAdLoader, mappingId: param.mappingId)
        loadCompletion = { (ad: NADNative?, error: Error?)  in
            guard let ad = ad, error == nil else {
                self.invokeDelegates(event: .onFailedToLoad, args: [self.KEY_ERROR_CODE: (error != nil ? error!._code : kNADNativeErrorCodeInvalidResponseType)])
                return
            }
            let hashCode = String(ad.hash)
            self.onConnect?(ad, hashCode)
            let bridgingAd = ["nativeAd": [
                "titleText": ad.shortText,
                "contentText": ad.longText,
                "promotionName": ad.promotionName,
                "promotionUrl": ad.promotionUrl,
                "adImageUrl": ad.imageUrl,
                "logoImageUrl": ad.logoUrl,
                "actionButtonText": ad.actionButtonText,
                "hashCode" : hashCode]
            ]
            self.invokeDelegates(event: .onLoaded, args: bridgingAd)
        }
    }
    
    override func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch targetMethod(from: call) {
        case .disableAutoReload:
            client.disableAutoReload()
        case .enableAutoReload:
            if let intervalMillis = NativeAdLoader.adUnit(from: call.arguments).intervalMillis {
                client.enableAutoReload(withInterval: TimeInterval(intervalMillis / 1000), completionBlock: loadCompletion)
                break
            }
            fallthrough
        case .loadAd:
            client.load(completionBlock: loadCompletion)
        default:
            super.handle(call, result: result)
            return
        }
        result(true)
    }
    
    class func adUnit(from argument: Any?) -> NativeAdUnitCodable {
        let jsonData = try! JSONSerialization.data(withJSONObject: argument!, options: [])
        return try! JSONDecoder().decode(NativeAdUnitCodable.self, from: jsonData)
    }
    
}

File: android/src/main/kotlin/net/nend/nendplugin/NativeAdLoader.kt
package net.nend.nendplugin

import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.PluginRegistry
import net.nend.android.NendAdNative
import net.nend.android.NendAdNativeClient
import net.nend.android.internal.connectors.NendNativeAdConnectorFactory

class NativeAdLoader(registrar: PluginRegistry.Registrar, arguments: Any) : AdBridger(registrar, NendPlugin.getAdUnitFrom(arguments).first) {
    lateinit var connectorCallback: NendPlugin.ConnectorCallback
    private val client : NendAdNativeClient
    private val adUnit = NendPlugin.getAdUnitFrom(arguments)

    private val callback = object : NendAdNativeClient.Callback {
        override fun onSuccess(nendAdNative: NendAdNative) {
            val connector = NendNativeAdConnectorFactory.createNativeAdConnector(nendAdNative)
            val hashCode = connector.hashCode().toString()
            connectorCallback.onConnect(connector, hashCode)
            val nativeAd = mapOf(
                    "titleText" to connector.shortText,
                    "contentText" to connector.longText,
                    "promotionName" to connector.promotionName,
                    "promotionUrl" to connector.promotionUrl,
                    "adImageUrl" to connector.adImageUrl,
                    "logoImageUrl" to connector.logoImageUrl,
                    "actionButtonText" to connector.actionButtonText,
                    "hashCode" to hashCode
            )
            invokeListenerEvent(CallbackName.OnLoaded, mapOf("nativeAd" to nativeAd))
        }

        override fun onFailure(nendError: NendAdNativeClient.NendError) {
            invokeListenerEvent(CallbackName.OnFailedToLoad, mapOf(KEY_ERROR_CODE to nendError.code))
        }
    }

    init {
        client = NendAdNativeClient(activity, adUnit.second, adUnit.third)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (getMethod(call, getTag(client))) {
            MethodName.LoadAd -> {
                client.loadAd(callback)
            }
            MethodName.EnableAutoReload -> {
                val interval = NendPlugin.simplePickFrom(call.arguments, "intervalMillis").toLong()
                client.enableAutoReload(interval, callback)
            }
            MethodName.DisableAutoReload -> {
                client.disableAutoReload()
            }
            else -> {
                super.onMethodCall(call, result)
                return
            }
        }
        result.success(true)
    }

    private fun invokeListenerEvent(event: CallbackName, args: Map<String, Any>?) {
        NendPlugin.channel.invokeMethod(getTag(client) + "." + event.toString().decapitalize(), mappingArguments(mappingId, args))
    }
}